### PR TITLE
[TBAA] Fix the case where a subobject gets accessed at a non-zero offset.

### PR DESCRIPTION
**Pull request by kosarev**

### CODE REVIEW

**Comment by kosarev** [comment]
@vedantparanjape-amd Vedant, can you confirm if this resolves https://github.com/llvm/llvm-project/issues/101164?

**Comment by vedantparanjape-amd** [comment]
> @vedantparanjape-amd Vedant, can you confirm if this resolves #101164?

Hello Ivan, sorry for the delay to get back. Let me verify this. It seems similar to my prototype fix !

**Review by antoniofrighetto — APPROVED** [review]
This makes sense to me, please wait for another reviewer's approval.

**Comment by kosarev** [comment]
Rebased. Ping.

File: llvm/lib/Analysis/TypeBasedAliasAnalysis.cpp
Comment (efriedma-quic) [inline llvm/lib/Analysis/TypeBasedAliasAnalysis.cpp:615]: I'm sorry, I can't follow the logic here at all.  What is the purpose of the code inside this if statement?

Comment (efriedma-quic) [reply llvm/lib/Analysis/TypeBasedAliasAnalysis.cpp:615]: I think I see... the old code detects whether both accesses refer to the same object, but we also care about overlapping objects.  According to the TBAA rules, objects have to be properly nested.  So the only way to get overlap is if one of the accesses is a "whole-object" access.  Therefore, this check.

LangRef says "the access type is a scalar type descriptor", which I guess isn't accurate?  Please propose a patch to fix the language there.

Comment (kosarev) [reply llvm/lib/Analysis/TypeBasedAliasAnalysis.cpp:615]: LangRef doesn't cover the new aggregate-enabled TBAA format, which this patch is about. The latest plan documentation-wise I'm aware of is to wait until the new format is mature enough to be enabled by default, and then update the documentation, see <https://reviews.llvm.org/D40975#955973>. Until then it's all work-in-progress and not something supposed to be exposed to the user.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**
